### PR TITLE
Feature/organ donor relationship

### PR DESCRIPTION
**Pull request by steffanie07**
This PR contains the following changes:
-  Creation of  Organ Table
-  Seeding/populating the organ table with data from existing form 
-  Creating relationship between donor and organs by creating a pivot/junction table called Donation Preferences
-  I have also added a Migrations.sql file that can be run by using:
`dotnet ef migrations script -o Migrations.sql`

The response from http:localhost:5059/api/GetDonors should look like :
 {
    "id": 12,
    "fullName": "Stephanie",
    "bloodType": "b+",
    "dateOfBirth": "2024-05-21T10:33:59.034",
    "contactNumber": "975222",
    "identityNumber": "0000a3",
    "residentialAddress": "Swieqi",
    "mailingAddress": "Swiwqi",
    "email": "minopbjboge@gmail.com",
    "telephoneNumber": "009192838",
    "mobileNumber": "009192838",
    "nationality": "Lebanese",
    "gender": "F",
    "preferredContact": "EMAIL",
    "donationPreferences": [
      3,
      5,
      9,
      10
    ]  
the donation preferences , hold the ids of the chosen organs during registration.

### CODE REVIEW

File: Donor/Controllers/DonorController.cs [hunk 1]
```diff
@@ -44,42 +45,48 @@ public string Ping()
...
-		}
+        /// Register a new donor
+        /// </summary>
+        /// <param name="donorDto"></param>
+        /// <returns></returns>
+        [Tags("Donor")]
+        [HttpPost("Register")]
+        public async Task<IActionResult> Register([FromBody] DonorDto donorDto)
+        {
+            if (donorDto == null)
```
Comment (josecarv) [inline Donor/Controllers/DonorController.cs]: This validation is not necessary, if the object is null the endpoint will never hit the endpoint.

File: Donor/Entities/DonationPreference.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+namespace Donor.Entities
+{
+    public class DonationPreference
```
Comment (josecarv) [inline Donor/Entities/DonationPreference.cs]: The relationship between Donor and Organ is a Many to Many relationship, if I'm not mistaken.
If that's the case the representation in .Net (EF 6.X) is through navigation properties.
**Therefore, in Donor you should have:** 
  public int  OrganId { get; set; }
  public ICollection<Organ> Organs{ get; set; }
In Organ you should have : 
  public int  DonorId { get; set; }
  public ICollection<Donor> Donors{ get; set; }

To sum up, the DonationPreference class is not needed as an entity in your code! But it is need it in SQL.

File: Donor/Entities/Donor.cs [hunk 3]
```diff
@@ -15,6 +19,13 @@ public class Donor
         public string MobileNumber { get; set; } = string.Empty;
         public string Nationality { get; set; } = string.Empty;
         public string Gender { get; set; } = string.Empty;
+        
+        public PreferredContactMode PreferredContact { get; set; }
+        public ICollection<DonationPreference> DonationPreferences { get; set; } = new List<DonationPreference>();
```
Comment (josecarv) [inline Donor/Entities/Donor.cs]: Instead of  a collection of DonationPreference
it should be  a collection of Organ plus the OrganId property.
This is related with my previous comment

File: Donor/Entities/Organ.cs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+namespace Donor.Entities
+{
+    public class Organ
+    {
+        public int Id { get; set; }
+        public string Name { get; set; } = string.Empty;
```
Comment (josecarv) [inline Donor/Entities/Organ.cs:6]: The relationship with the Donor is missing.
This class should have a collection of Donor and DonorId property

File: Donor/Models/DonationPreferenceDto.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+namespace Donor.Models
+{
+    public class DonationPreferenceDto
```
Comment (josecarv) [inline Donor/Models/DonationPreferenceDto.cs]: The representation of this class is no need it

File: Donor/Profiles/DonorsProfile.cs [hunk 6]
```diff
@@ -1,12 +1,20 @@
...
 
 namespace Donor.Profiles
 {
     public class DonorsProfile : Profile
     {
         public DonorsProfile()
         {
-            CreateMap<Models.DonorDto, Entities.Donor>();
+          CreateMap<DonorDto, Entities.Donor>()
+                .ForMember(dest => dest.DonationPreferences, opt => opt.Ignore());
```
Comment (josecarv) [inline Donor/Profiles/DonorsProfile.cs]: This mapping is going to change base on the relationship between Donor and Organ classes

File: Donor/Repositories/DonorRepository.cs [hunk 7]
```diff
@@ -1,80 +1,105 @@
...
         {
             _context = context ?? throw new ArgumentNullException(nameof(context));
-			_log = log ?? throw new ArgumentNullException(nameof(log));
-		}
-	
+            _log = log ?? throw new ArgumentNullException(nameof(log));
+        }
 
-        public async Task AddDonorAsync(Entities.Donor donor)
+        public async Task AddDonorAsync(Entities.Donor donor, List<int> donationPreferences)
```
Comment (josecarv) [inline Donor/Repositories/DonorRepository.cs]: To add a Donor you should only need a Donor! no the donationPreferences list

File: Donor/Repositories/DonorRepository.cs [hunk 8]
```diff
@@ -1,80 +1,105 @@
...
         {
-			try
-			{
-				await _context.Donors.AddAsync(donor);				
+            using var transaction = await _context.Database.BeginTransactionAsync();
+            try
+            {
+                await _context.Donors.AddAsync(donor);
+                await _context.SaveChangesAsync();
```
Comment (josecarv) [inline Donor/Repositories/DonorRepository.cs:34]: We should try to make one roundtrip to the DB per operation.
This is One roundtrip to add the Donor which is good !

File: Donor/Repositories/DonorRepository.cs [hunk 9]
```diff
@@ -1,80 +1,105 @@
...
+                if (donationPreferences != null && donationPreferences.Any())
+                {
+                    foreach (var organId in donationPreferences)
+                    {
+                        var preference = new DonationPreference
+                        {
+                            DonorId = donor.Id,
+                            OrganId = organId
+                        };
+                        await _context.DonationPreferences.AddAsync(preference);
```
Comment (josecarv) [inline Donor/Repositories/DonorRepository.cs]: But in the same operation you are making a second roundtrip to the DB.
Which is a performance issue !

File: Donor/Repositories/DonorRepository.cs [hunk 10]
```diff
@@ -1,80 +1,105 @@
...
-			{
-
-				_log.LogError(ex, "Error getting all donors");
-				throw;
-			}
+            IEnumerable<Entities.Donor> donors = new List<Donor.Entities.Donor>();
+            try
+            {
+                donors = await _context.Donors
+                    .Include(d => d.DonationPreferences)
```
Comment (josecarv) [inline Donor/Repositories/DonorRepository.cs]: .Include(d => d.DonationPreferences) will not be necessary when you make the correction on the relationship between Donor and Organ

File: Donor/Repositories/IDonorRepository.cs [hunk 11]
```diff
@@ -1,10 +1,12 @@
 ﻿using System.Threading.Tasks;
+using Donor.Models;
+
 
 namespace Donor.Repositories
 {
     public interface IDonorRepository
     {
-        Task AddDonorAsync(Entities.Donor donor);
+        Task AddDonorAsync(Entities.Donor donor,List<int> donationPreferences);
```
Comment (josecarv) [inline Donor/Repositories/IDonorRepository.cs]: The second parameter is not need it

**Review by josecarv — CHANGES_REQUESTED** [review]
Please take a look at the comments and make the changes.
